File: Projection/customevideoplayer/src/main/java/com/chen/videoplayer/player/VideoPlayerController.java
package com.chen.videoplayer.player;

import android.content.Context;
import android.os.Handler;
import android.os.Message;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.bumptech.glide.Glide;
import com.chen.videoplayer.R;

/**
 * Created by Administrator on 2017/6/5.
 * 视频播放控制器
 */
public class VideoPlayerController extends FrameLayout implements
        View.OnClickListener,
        SeekBar.OnSeekBarChangeListener,
        View.OnTouchListener{

    private VideoPlayerControl mVideoPlayerControl;

    private Context mContext;

    private ImageView mImage; // 封面图
    private ImageView mCenterStart;  // 中间播放按钮
    private LinearLayout mTop;  // 顶部布局
    private ImageView mBack;   // 顶部布局左上角返回按键
    private TextView mTitle;   // 视频标题
    private LinearLayout mBottom;   // 底部布局
    private ImageView mRestartPause; // 底部布局 左下角 播放、暂停控件
    private TextView mPosition;   // 当前播放时间点
    private TextView mDuration;   // 视频总时长
    private SeekBar mSeek;     // 视频播放进度拖动条
    private ImageView mFullScreen;  // 横竖屏切换控件
    private LinearLayout mLoading;  // 加载视频的布局
    private TextView mLoadText;   // 加载视频布局的文案
    private LinearLayout mError;  // 视频加载错误的布局
    private TextView mRetry;      // 点击重试
    private LinearLayout mCompleted;  // 视频播放完成现实的布局
    private TextView mReplay;  // 重新播放
    private TextView mShare;   // 分享

    private boolean topBottomVisible = false;

    private String mUrl;
    private int screenWidth;
    private int screenHeight;

    public VideoPlayerController(@NonNull Context context) {
        super(context);
        init(context);
    }

    private void init(Context context) {
        inflate(context, R.layout.video_palyer_controller, this);
        mContext = context;
        screenWidth = VideoPlayerUtil.getScreenWidth(context);
        screenHeight = VideoPlayerUtil.getScreenHeight(context);

        mCenterStart = (ImageView) findViewById(R.id.center_start);
        mImage = (ImageView) findViewById(R.id.image);

        mTop = (LinearLayout) findViewById(R.id.top);
        mBack = (ImageView) findViewById(R.id.back);
        mTitle = (TextView) findViewById(R.id.title);

        mBottom = (LinearLayout) findViewById(R.id.bottom);
        mRestartPause = (ImageView) findViewById(R.id.restart_or_pause);
        mPosition = (TextView) findViewById(R.id.position);
        mDuration = (TextView) findViewById(R.id.duration);
        mSeek = (SeekBar) findViewById(R.id.seek);
        mFullScreen = (ImageView) findViewById(R.id.full_screen);

        mLoading = (LinearLayout) findViewById(R.id.loading);
        mLoadText = (TextView) findViewById(R.id.load_text);

        mError = (LinearLayout) findViewById(R.id.error);
        mRetry = (TextView) findViewById(R.id.retry);

        mCompleted = (LinearLayout) findViewById(R.id.completed);
        mReplay = (TextView) findViewById(R.id.replay);
        mShare = (TextView) findViewById(R.id.share);

        mCenterStart.setOnClickListener(this);
        this.setOnClickListener(this);
        mReplay.setOnClickListener(this);
        mSeek.setOnSeekBarChangeListener(this);
        mRestartPause.setOnClickListener(this);
        mFullScreen.setOnClickListener(this);
        mBack.setOnClickListener(this);
        setOnTouchListener(this);
    }

    public void setVideoPlayer(VideoPlayerControl videoPlayerControl){
        mVideoPlayerControl = videoPlayerControl;
    }

    public void setTitle(String title){
        mTitle.setText(title);
    }

    public void setImage(String url){
        mUrl = url;
        Glide.with(mContext)
                .load(url)
                .placeholder(R.drawable.img_default)
                .into(mImage);
    }

    @Override
    public void onClick(View v) {
        if(v == mCenterStart){
//            if(mVideoPlayerControl.isIdle()){
                mVideoPlayerControl.start();
//            }
        }else if(v == this){
            if(mVideoPlayerControl.isPlaying()
                    || mVideoPlayerControl.isPaused()
                    || mVideoPlayerControl.isBufferingPlaying()
                    || mVideoPlayerControl.isBufferingPaused()){
                setTopBottomVisible(!topBottomVisible);
            }
        }else if(v == mReplay){
            mVideoPlayerControl.release();
            mVideoPlayerControl.start();
            mCompleted.setVisibility(GONE);
        }else if(v == mShare){
            Toast.makeText(mContext, "分享", Toast.LENGTH_SHORT).show();
        }else if(v == mRestartPause){  //暂停或者播放
            if(mVideoPlayerControl.isPlaying() || mVideoPlayerControl.isBufferingPlaying()){
                mVideoPlayerControl.pause();
            }else if (mVideoPlayerControl.isPaused() || mVideoPlayerControl.isBufferingPaused()){
                mVideoPlayerControl.restart();
            }
        }else if(v == mFullScreen){
            if(mVideoPlayerControl.isNormalScreen()){
                mVideoPlayerControl.enterFullScreen();
            }else if(mVideoPlayerControl.isFullScreen()){
                mVideoPlayerControl.exitFullScreen();
            }
        }else if(v == mBack){
            if(mVideoPlayerControl.isFullScreen()){
                mVideoPlayerControl.exitFullScreen();
            }
        }
    }

    /**
     * 设置头部和底部布局是否隐藏
     */
    private void setTopBottomVisible(boolean topBottomVisible) {
        mTop.setVisibility(topBottomVisible ? VISIBLE : INVISIBLE);
        mBottom.setVisibility(topBottomVisible ? VISIBLE : INVISIBLE);
        this.topBottomVisible = topBottomVisible;

        //3秒显示还没有关闭就自动关闭
        if(topBottomVisible){
            postDelayed(new Runnable() {
                @Override
                public void run() {
                    mTop.setVisibility(INVISIBLE);
                    mBottom.setVisibility(INVISIBLE);
                    VideoPlayerController.this.topBottomVisible = false;
                }
            }, 8000);
        }
    }

    /**
     * 设置播放器工作状态
     */
    public void setControllerState(int playState, int windowState){
        switch (playState){
            case VideoPlayer.STATE_IDLE:
                mBottom.setVisibility(GONE);
                break;
            case VideoPlayer.STATE_PREPARING:
                mImage.setVisibility(GONE);
                mLoading.setVisibility(VISIBLE);
                mLoadText.setText("正在准备...");
                mCenterStart.setVisibility(GONE);
                mBack.setVisibility(GONE);
                mTop.setVisibility(GONE);
                mBottom.setVisibility(GONE);
                break;
            case VideoPlayer.STATE_PREPARED: // 播放准备就绪
                startUpdateProgress();
                break;
            case VideoPlayer.STATE_BUFFERING_PLAYING: //播放时缓冲
                mLoading.setVisibility(VISIBLE);
                mLoadText.setText("正在缓冲");
                mRestartPause.setImageResource(R.drawable.ic_player_pause);
                break;
            case VideoPlayer.STATE_BUFFERING_PAUSED:  //暂停时缓冲
                mLoading.setVisibility(VISIBLE);
                mLoadText.setText("正在缓冲");
                mRestartPause.setImageResource(R.drawable.ic_player_start);
                break;
            case VideoPlayer.STATE_PLAYING:
                mLoading.setVisibility(GONE);
                mRestartPause.setImageResource(R.drawable.ic_player_pause);
                break;
            case VideoPlayer.STATE_PAUSED:
                mLoading.setVisibility(GONE);
                mRestartPause.setImageResource(R.drawable.ic_player_start);
                break;
            case VideoPlayer.STATE_COMPLETED:
                cancelUpdateProgress();
                mCompleted.setVisibility(VISIBLE);
                mImage.setVisibility(VISIBLE);
                break;
        }

        switch (windowState){
            case VideoPlayer.PLAYER_FULL_SCREEN:
                mBack.setVisibility(VISIBLE);
                mFullScreen.setVisibility(VISIBLE);
                mFullScreen.setImageResource(R.drawable.ic_player_shrink);
                break;
            case VideoPlayer.PLAYER_NORMAL:
                mBack.setVisibility(GONE);
                mFullScreen.setVisibility(VISIBLE);
                mFullScreen.setImageResource(R.drawable.ic_player_enlarge);
                break;
            case VideoPlayer.PLAYER_TINY_WINDOW:
                mBack.setVisibility(GONE);
                mFullScreen.setVisibility(GONE);
                break;
        }
    }

    private final static int MSG_ID = 0x01;
    private Handler handler = new Handler(){
        @Override
        public void handleMessage(Message msg) {
            updateProgress();
            handler.sendEmptyMessageDelayed(MSG_ID, 300);
        }
    };
    /**
     * 更新进度
     */
    private void startUpdateProgress(){
        handler.sendEmptyMessageDelayed(MSG_ID, 300);
    }

    /**
     * 取消更新
     */
    private void cancelUpdateProgress(){
        if(handler != null){
            handler.removeMessages(MSG_ID);
        }
    }

    private void updateProgress(){
        int duration = mVideoPlayerControl.getDuration();
        int currentPosition = mVideoPlayerControl.getCurrentProgress();
        mSeek.setSecondaryProgress(mVideoPlayerControl.getBufferPercent());
        mSeek.setProgress((int) (currentPosition*1.0f/duration*100));
        mPosition.setText(VideoPlayerUtil.formatTime(currentPosition));
        mDuration.setText(VideoPlayerUtil.formatTime(duration));
    }

    /**
     *  暂停播放器
     */
    public void onPause(){

        if(mVideoPlayerControl.isIdle()) return;

        mVideoPlayerControl.pause();
    }

    /**
     *  播放器继续播放
     */
    public void onRestart(){

        if(mVideoPlayerControl.isIdle()) return;   //播放器还只是空闲状态就不继续播放视频了

        mVideoPlayerControl.restart();
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
        cancelUpdateProgress();
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        int progress = (int) (mVideoPlayerControl.getDuration() * seekBar.getProgress() / 100f);
        mVideoPlayerControl.seekTo(progress);
        startUpdateProgress();
    }

    public void reset() {
        mError.setVisibility(GONE);
        mLoading.setVisibility(GONE);
        mCompleted.setVisibility(GONE);
    }

    private float startX;
    private float startY;
    @Override
    public boolean onTouch(View v, MotionEvent event) {

        // 不是小窗口
        if(!mVideoPlayerControl.isTinyScreen()) return super.onTouchEvent(event);

        // 是小窗口  拖动小窗口的代码
        switch (event.getAction()){
            case MotionEvent.ACTION_DOWN:
                startX = event.getRawX();
                startY = event.getRawY();
                break;
            case MotionEvent.ACTION_MOVE:

                float endX = event.getRawX();
                float endY = event.getRawY();

                float dx = endX - startX;
                float dy = endY - startY;

                LayoutParams params = (LayoutParams) mVideoPlayerControl.getContainer().getLayoutParams();

                int left = (int) (params.leftMargin + dx);
                int top = (int) (params.topMargin + dy);
                int viewHeight = mVideoPlayerControl.getContainer().getHeight() + 50;
                int viewWidth = mVideoPlayerControl.getContainer().getWidth();

                if(left < -1){
                    left = 0;
                }else if(left > screenWidth - viewWidth){
                    left = screenWidth - viewWidth;
                }

                if(top < -1){
                    top = 0;
                }else if(top > screenHeight - viewHeight){
                    top = screenHeight - viewHeight;
                }

                params.leftMargin = left;
                params.topMargin = top;
                mVideoPlayerControl.getContainer().setLayoutParams(params);

                startX = endX;
                startY = endY;
                break;
        }

        return super.onTouchEvent(event);
    }
}

File: Projection/test/src/main/java/com/kuanquan/test/controller/VideoPlayerController.kt
package com.kuanquan.test.controller

import android.content.Context
import android.content.pm.ActivityInfo
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.*
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.OnLifecycleEvent
import com.kuanquan.test.R
import com.kuanquan.test.player.SuperPlayer
import com.kuanquan.test.player.SuperPlayerDef
import com.kuanquan.test.player.SuperPlayerImpl
import com.kuanquan.test.player.SuperPlayerObserver
import com.kuanquan.test.projection.DLNAUtils
import com.kuanquan.test.seek.VerticalSeekBar
import com.tencent.rtmp.ui.TXCloudVideoView

/**
 * 播放器界面上的控制面板布局
 */
class VideoPlayerController : FrameLayout, View.OnClickListener, OnSeekBarChangeListener, LifecycleObserver {

    constructor(context: Context) : super(context) {
        initView(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initView(context)
    }

    private var mCenterStart: ImageView? = null // 中间播放按钮
    private var verticalSeekBar: VerticalSeekBar? = null // 垂直进度条，调整声音
    private var mTop: LinearLayout? = null // 顶部布局
    private var mBack: ImageView? = null // 顶部布局左上角返回按键
    private var mTitle: TextView? = null // 视频标题
    private var mBottom: LinearLayout? = null // 底部布局
    private var mRestartPause: ImageView? = null // 底部布局 左下角 播放、暂停控件
    private var mPositionTv: TextView? = null // 当前播放时间点
    private var mDurationTv: TextView? = null // 视频总时长
    private var mSeekBar: SeekBar? = null // 视频播放进度拖动条
    private var voiceIv: ImageView? = null // 播放器音量调节按钮
    private var mFullScreen: ImageView? = null // 横竖屏切换控件
    private var mLoading: LinearLayout? = null // 加载视频的布局
    private var mLoadText: TextView? = null // 加载视频布局的文案
    private var mError: LinearLayout? = null // 视频加载错误的布局
    private var mRetry: TextView? = null // 点击重试
    private var mCompleted: LinearLayout? = null // 视频播放完成现实的布局
    private var mReplay: TextView? = null // 重新播放
    private var mShare: TextView? = null // 分享

    private val screenWidth: Int by lazy { context.resources.displayMetrics.widthPixels }

    private val screenHeight: Int by lazy { context.resources.displayMetrics.heightPixels }

    /**
     * 把自定义 View 的生命周期和所在页面的 Activity 或 Fragment 的生命周期关联起来
     */
    fun addLifecycleObserver(lifecycleOwner: LifecycleOwner?){
        lifecycleOwner?.lifecycle?.addObserver(this)
    }

    private fun initView(context: Context) {
        View.inflate(context, R.layout.video_palyer_controller, this)

        verticalSeekBar = findViewById(R.id.vertical_Seekbar)
        initVerticalSeekBar(verticalSeekBar)
        mCenterStart = findViewById(R.id.center_start)

        mTop = findViewById(R.id.top)
        mBack = findViewById(R.id.back)
        mTitle = findViewById(R.id.title)

        mBottom = findViewById(R.id.bottom)
        mRestartPause = findViewById(R.id.restart_or_pause)
        mPositionTv = findViewById(R.id.position)
        mDurationTv = findViewById(R.id.duration)
        mSeekBar = findViewById(R.id.seek)
        voiceIv = findViewById(R.id.voice_iv)
        mFullScreen = findViewById(R.id.full_screen)

        mLoading = findViewById(R.id.loading)
        mLoadText = findViewById(R.id.load_text)

        mError = findViewById(R.id.error)
        mRetry = findViewById(R.id.retry)

        mCompleted = findViewById(R.id.completed)
        mReplay = findViewById(R.id.replay)
        mShare = findViewById(R.id.share)

        mCenterStart?.setOnClickListener(this)
        setOnClickListener(this)
        mReplay?.setOnClickListener(this)
        mRestartPause?.setOnClickListener(this)
        mFullScreen?.setOnClickListener(this)
        mBack?.setOnClickListener(this)
        mShare?.setOnClickListener(this)

        mSeekBar?.setOnSeekBarChangeListener(this)
    }

    private fun initVerticalSeekBar(verticalSeekBar: VerticalSeekBar?) {
        verticalSeekBar?.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onStartTrackingTouch(seekBar: SeekBar) {
                handler.removeCallbacksAndMessages(null)
            }

            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                mSuperPlayer?.setAudioPlayoutVolume(progress)
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                setTopBottomVisible(true)
            }
        })
    }

    private var videoUrl: String? = null
    private var topBottomVisible = false // 播放器底部和顶部状态栏隐藏和显示
    private var mSuperPlayer: SuperPlayer? = null // 播放器常用操作接口类

    /**
     * 设置腾讯云播放器
     */
    fun semTXCloudVideoView(tXCloudVideoView: TXCloudVideoView) {
        setPlayer(SuperPlayerImpl(context, tXCloudVideoView))
    }

    /**
     * 设置播放器回掉
     */
    private fun setPlayer(player: SuperPlayer) {
        mSuperPlayer = player
        mSuperPlayer?.setObserver(mSuperPlayerObserver)
    }

    /**
     * 设置播放URL
     */
    fun setPlayUrl(url: String) {
        videoUrl = url
    }

    fun setTitle(title: String?) {
        mTitle?.text = title
    }

    /**
     * 恢复播放
     */
    fun resume(){
        mSuperPlayer?.resume()
    }

    /**
     * 开始播放
     */
    fun startPlayer(){
        mSuperPlayer?.play(videoUrl)
    }

    /**
     * 暂停播放
     */
    fun pausePlayer(){
        mSuperPlayer?.pause()
    }

    /**
     * 停止播放
     */
    fun stopPlayer(){
        mSuperPlayer?.stop()
    }

    /**
     * 硬解
     */
    fun enableHardwareDecode(){
        mSuperPlayer?.enableHardwareDecode(true)
    }

    /**
     * 隐藏播放器中间的播放按钮
     */
   fun hideView(){
       mCenterStart?.visibility = View.GONE
    }

    /**
     * 指定播放的位置
     */
    fun setSeekJump(jDuration: Long){
        jumpDuration = jDuration
    }

    /**
     * 指定播放的位置
     */
    fun setSeek(jDuration: Long){
        mSuperPlayer?.seek(jDuration.toInt())
    }

    private var totalDuration = 0L // 视屏总时长
    private var currentDuration = 0L // 当前播放的时间点
    private var jumpDuration = 0L // 跳转到指定地方播放

    /**
     * 播放器状态回掉
     */
    private val mSuperPlayerObserver = object : SuperPlayerObserver() {
        override fun onPlayProgress(current: Long, duration: Long) {
            super.onPlayProgress(current, duration)
            totalDuration = duration
            currentDuration = current
            mPositionTv?.text = DLNAUtils.formattedTime(current)
            mDurationTv?.text = DLNAUtils.formattedTime(duration)

            if (totalDuration > 0) {
                mSeekBar?.progress = (current * 100 / totalDuration).toInt()
            }
        }

        override fun onPlayStart(name: String?) {
            super.onPlayStart(name)
            Log.e("VideoPlayerController", "onPlayStart -》 onPlayStart ")
            if (jumpDuration > 0) {
                mSuperPlayer?.seek(jumpDuration.toInt())
                mSuperPlayer?.resume()
            }
            mCompleted?.visibility = View.GONE
            mRestartPause?.setImageResource(R.drawable.ic_player_pause)
        }

        override fun onPlayBegin(name: String?) {
            super.onPlayBegin(name)
            Log.e("VideoPlayerController", "onPlayBegin -》 onPlayBegin ")
            mCompleted?.visibility = View.GONE
            mRestartPause?.setImageResource(R.drawable.ic_player_pause)
        }

        override fun onPlayPause() {
            super.onPlayPause()
            mRestartPause?.setImageResource(R.drawable.ic_player_start)
        }

        override fun onPlayStop() {
            super.onPlayStop()
            mCompleted?.visibility = View.VISIBLE
        }
    }

    override fun onClick(v: View?) {
        when (v) {
            this -> { // 点击隐藏和显示顶部和底部导航栏用的
                if (mSuperPlayer?.playerState == SuperPlayerDef.PlayerState.PLAYING || mSuperPlayer?.playerState == SuperPlayerDef.PlayerState.PAUSE) {
                    setTopBottomVisible(!topBottomVisible)
                }
            }
            mCenterStart -> { // 播放器中间的播放按钮
                mSuperPlayer?.play(videoUrl)
                mCenterStart?.visibility = View.GONE
            }
            mReplay -> { // 重新播放
                mCompleted?.visibility = View.GONE
                mSuperPlayer?.reStart()
            }
            mShare -> {
                Toast.makeText(context, "分享", Toast.LENGTH_SHORT).show()
            }
            mRestartPause -> {  // 暂停或者播放
                if (mSuperPlayer?.playerState == SuperPlayerDef.PlayerState.PLAYING) {
                    mSuperPlayer?.pause()
                } else if (mSuperPlayer?.playerState == SuperPlayerDef.PlayerState.PAUSE) {
                    mSuperPlayer?.resume()
                }
            }
            mBack -> {
                mPageFinishListener?.onScreenIcon(currentDuration)
            }
            mFullScreen -> {
                mPageFinishListener?.onScreenIcon(0)
            }
        }
    }

    override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {

    }

    override fun onStartTrackingTouch(seekBar: SeekBar?) {
        handler.removeCallbacksAndMessages(null)
    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {
        setTopBottomVisible(true)
        val currentTime = seekBar.progress * totalDuration / 100
        mSuperPlayer?.seek(currentTime.toInt())
        mSuperPlayer?.resume()
    }

    private fun reset() {
        mError?.visibility = View.GONE
        mLoading?.visibility = View.GONE
        mCompleted?.visibility = View.GONE
    }

    /**
     * 设置头部和底部布局是否隐藏
     */
    private fun setTopBottomVisible(topBottomVisible: Boolean) {
        mTop?.visibility = if (topBottomVisible) View.VISIBLE else View.INVISIBLE
        mBottom?.visibility = if (topBottomVisible) View.VISIBLE else View.INVISIBLE
        verticalSeekBar?.visibility = if (topBottomVisible) View.VISIBLE else View.INVISIBLE
        this.topBottomVisible = topBottomVisible

        // 8秒显示还没有关闭就自动关闭
        if (topBottomVisible) {
            handler.postDelayed({
                mTop?.visibility = View.INVISIBLE
                mBottom?.visibility = View.INVISIBLE
                verticalSeekBar?.visibility = View.INVISIBLE
                this.topBottomVisible = false
            }, 8000)
        } else {
            handler.removeCallbacksAndMessages(null)
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroy() {
        mSuperPlayer?.stop()
        mSuperPlayer?.destroy()
        reset()
        handler.removeCallbacksAndMessages(null)
    }

    fun setPageFinish(listener: PageFinishListener?) {
        mPageFinishListener = listener
    }

    private var mPageFinishListener: PageFinishListener? = null
    interface PageFinishListener {
        fun onScreenIcon(currentDuration: Long)
    }

}